eip-0008.md

```diff
@@ -0,0 +1,479 @@
+
+##### Footnotes
+1) Cost estimation is a part of the existing proof generation in Ergo, but AOT costing
+is used instead of JIT costing in version v1 of Ergo protocol.
+
+2) A default `costPrice` can be used to make this register optional. The same is true for
+other registers of `feeOut` box. This makes the new protocol compatible with existing
+applications which send simple pay-to-pk address transactions.
+
+3) The protocol penalize senders of the invalid transactions as long as the
+transactions passes all intrinsic validation checks. This means the sender should be
+charged if the transaction is invalidated due to scripts verification, which may be expensive. 
+The sender should ensure, that the transaction inputs scripts can be verified in
+any context where the transaction is intrinsically valid. For example tx can be
+intrinsically invalid if it spends a box which has already been spent. The sender can
+limit intrinsic validity of the transaction by specifying R6 register of `feeOut` box with 
+valid height interval.
+
+#### The Benefits of the JIT costing
+
+1) A miner will never execute more operations for tx validation than it is payed in
+miner's fee outputs, thus block validation costs are strictly limited by sum of fee
+outputs and the miner's validation expenses are fully rewarded, no matter how complex the
+script operations could be. This will enable easy long-term evolution of ErgoTree language
+via soft-forks.
+
+2) When selecting transactions from the pool, a miner can quickly filter out transactions
+that doesn't fit into MaxBlockCost using `feeLimit / costPrice`.
+
+3) In many cases (such as P2PK contracts and others) when the cost of transaction
+validation doesn't depend on the context the cost of `tx` may be exactly predicted by the
+sender. In such cases:
+      - no refund outputs will be created for the transactions spending such "simple"
+      boxes. Thus, the protocol doesn't have overhead in simple cases.
+      - existing applications can be supported (without change) by using reasonable
+      default values when the registers of `feeOut` box are not present.
+
+4) The context independence of scripts and hence the exact prediction of `actCost` is a
+static property of ErgoTree, it can be determined by analysing ErgoTree for which Context
+properties the script uses. This can be controlled and enforced by the sender. Moreover,
+it is possible to extend this property to a wider range of scripts by introducing a
+_validity interval_ of a transaction (see the `R6` register of the `feeOut` box). This is
+an interval of the blockchain heights during which the transaction can be processed. Thus,
+any scripts which run during validation can assume that the current height is within that
+interval, but do not rely on the precise value of the current height. This will make
+`actCost` exactly predictable even for scripts what have conditions on `HEIGHT` (like
+`HEIGHT > deadline`).
```

Comment:
Why to do additional changes in the protocol for HEIGHT context-independence only ? In examples we have dependency on a spending transaction is more critical.